runtime/src/main/java/org/corfudb/protocols/logprotocol/MultiObjectSMREntry.java

```diff
@@ -81,40 +80,75 @@ public void mergeInto(MultiObjectSMREntry other) {
+            byte[] streamUpdates = new byte[updateLength];
+            b.readBytes(streamUpdates);
+            streamBuffers.put(streamId, streamUpdates);
         }
     }
 
     @Override
     public void serialize(ByteBuf b) {
         super.serialize(b);
-        b.writeInt(entryMap.size());
-        entryMap.entrySet().stream()
+        b.writeInt(streamUpdates.size());
+        streamUpdates.entrySet().stream()
                 .forEach(x -> {
+                    // Stream payload structure
+                    // | stream id | payload size | SMR Update 1 | ... | SMR Update N|
                     b.writeLong(x.getKey().getMostSignificantBits());
                     b.writeLong(x.getKey().getLeastSignificantBits());
-                    Serializers.CORFU.serialize(x.getValue(), b);
+                    int payloadSizeIndex = b.writerIndex();
+                    b.writeInt(0);
+                    int payloadIndex = b.writerIndex();
+                    b.writeInt(x.getValue().getUpdates().size());
+                    x.getValue().getUpdates().stream().forEach(smrEntry -> Serializers.CORFU.serialize(smrEntry, b));
+                    int length = b.writerIndex() - payloadIndex;
+                    b.writerIndex(payloadSizeIndex);
+                    b.writeInt(length);
+                    b.writerIndex(payloadIndex + length);
                 });
     }
 
     /**
      * Get the list of SMR updates for a particular object.
+     *
      * @param id StreamID
      * @return an empty list if object has no updates; a list of updates if exists
      */
     @Override
     public List<SMREntry> getSMRUpdates(UUID id) {
-        MultiSMREntry entry = entryMap.get(id);
-        return entryMap.get(id) == null ? Collections.emptyList() :
-                entry.getUpdates();
+
+        // Since a stream buffer should only be deserialized once and multiple
+        // readers can deserialize different stream updates within the same container,
+        // synchronization on a per-stream basis is required.
+        MultiSMREntry resMultiSmrEntry = streamUpdates.computeIfAbsent(id, k -> {
+
+            if (streamBuffers.containsKey(id)) {
+                // The stream exists and it needs to be deserialized
+                byte[] streamUpdatesBuf = streamBuffers.get(id);
+                ByteBuf buf = Unpooled.wrappedBuffer(streamUpdatesBuf);
+                int numUpdates = buf.readInt();
+                List<SMREntry> smrEntries = new ArrayList<>(numUpdates);
+                for (int update = 0; update < numUpdates; update++) {
+                    smrEntries.add((SMREntry) Serializers.CORFU.deserialize(buf, null));
+                }
+
+                MultiSMREntry multiSMREntry = new MultiSMREntry(smrEntries);
+                multiSMREntry.setGlobalAddress(getGlobalAddress());
+                streamBuffers.remove(id);
+                return multiSMREntry;
+            } else {
+                return null;
```

Comment:
Can we flip the `if` logic and remove `else`?

Reply:
Done.